blog/_posts/2019-09-06-profilers.md

```diff
@@ -0,0 +1,405 @@
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    1 int.jl:52; -
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎  1  int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎  1  int.jl:52; -
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎   1  int.jl:52; -
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎ 1  int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎ 4  int.jl:52; -
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    5   int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    2   int.jl:52; -
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎  2   int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎  2   int.jl:52; -
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎     1   int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎   1   int.jl:52; -
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎ 1   int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    1   int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎  2   int.jl:53; +
+```
+
+We also see this has issues with identifying the stack in many cases (almost 40% failed!), similar to `perf`. One way to improve that (~10% failure rate) is to disable the omit-frame-pointer option, like we did before. Currently this requires this source code change:
+
+```diff
+diff --git a/src/jitlayers.cpp b/src/jitlayers.cpp
+index cf339d8a4b..f6e9c08e2c 100644
+--- a/src/jitlayers.cpp
++++ b/src/jitlayers.cpp
+@@ -3,9 +3,7 @@
+ #include "llvm-version.h"
+ #include "platform.h"
+ #include "options.h"
+-#if defined(_OS_WINDOWS_) || defined(_OS_FREEBSD_)
+ #  define JL_DISABLE_FPO
+-#endif
+ 
+ #include <iostream>
+ #include <sstream>
+```
+
+But otherwise, this seems fairly accurate, but yikes, what an arrow! I needed over 40 frames this way, and that makes this pretty hard to interpret.
+Spiky frame-graphs like this can be quite misleading to look at since the visual size (height) isn't relevant to performance.
+And it make it really hard to see where the time is being spent in aggregate.
+
+Julia implements a `noisefloor` option as well as `mincount` and `maxdepth` to help with some situations, but they do nothing for this example.
+The `mincount` option (also present in `Instruments.app` for min and max # samples; and in `perf` as a min percent threshold) drops any line that doesn't meet that threshold. The `maxdepth` option (aka `print_limit` in `perf`) cuts off the tree after some recursion limit.
+The `noisefloor` option is similar to `mincount`, but operates as a relative fraction of the parent frame instead of a constant threshold (I'm not aware of a similar option in other tools). This can help hide frames from the tree that contribute little to the immediate performance. This is based on the value engineering principle of focusing on the largest contributors first. For many problems, these can be greatly beneficial.
```

Comment:
2nd mention of `noisefloor`, would be better to consolidate. This explanation also suggests that it's a constant fraction, but when I last worked on this it tested whether the child was more than `noisefloor*sqrt(parent)` samples, as that's the fluctuation expected from sampling noise.